### DISCUSSION TITLE
Modding without "round-tripping"

### DISCUSSION THREAD

**Discussion by RenateUSB** [iBotPeaches/Apktool#2947] [General]
Don't get me wrong, I love apktool.

Still, to mod a 30MB framework.jar using Android 10 hidden API flags has not been possible recently because of a problem which will shortly be fixed. In the interim I had to come up with a solution.

First, do an `apktool d` to see what/where you have to mod. Then get the address of the method and instruction that you need to change. I'm not sure what tools are out there to get the actual file offset. I hacked something for that. Now your mod consists of a few well placed smali instructions directly into the dex. Of course for extensive changes this would be very complicated. You're limited by the original size of the method and fixing try/catches would be painful. Still, many mods are just turning a method into return(true) or not taking a branch.

The best part is that now your mod takes a fraction of a second and that nothing has been changed except what is necessary. Of course you still need apktool for analysis of both the original and the modded as a check.

**Comment by iBotPeaches** [comment]
I'm not sure what you are suggesting though.

Without round tripping is basically Xposed no? You have to round trip disassemble/assemble even if you used smali/baksmali directly without Apktool.

Or perhaps you are looking for something like DexPatcher - https://github.com/DexPatcher/dexpatcher-tool

**Comment by RenateUSB** [comment]
No, that DexPatcher and Xposed are interesting, but not what I'm talking about.
I'm talking about modding without doing any `apktool b` (or equivalent).
Just take a dex and modify it at some file offset. Uncompressed zipping is all you have to do.

When a new version of the jar/apk comes out you don't even have to `apktool d` it.
You get the new method address, add a known offset, check it, patch it, done.
